drivers/kscan/kscan_matrix.c

```diff
@@ -0,0 +1,353 @@
+					sys_clock_timeout_end_calc(K_USEC(data->poll_timeout));
+
+			} else if (k_uptime_ticks() > poll_time_end) {
+				break;
+			}
+
+			/* Subtract the time invested from the sleep period
+			 * in order to compensate for the time invested
+			 * in debouncing a key
+			 */
+			current_cycles = k_cycle_get_32();
+			cycles_diff = current_cycles - start_period_cycles;
+			wait_period = data->poll_period - k_cyc_to_us_floor32(cycles_diff);
+
+			/* Override wait_period in case it is less than 1 ms */
+			if (wait_period < USEC_PER_MSEC) {
+				wait_period = USEC_PER_MSEC;
+			}
+
+			/* wait period results in a larger number when
+			 * current cycles counter wrap. In this case, the
+			 * whole poll period is used
+			 */
+			if (wait_period > data->poll_period) {
+				LOG_DBG("wait_period : %u", wait_period);
+
+				wait_period = data->poll_period;
+			}
+
+			/* Allow other threads to run while we sleep */
+			k_usleep(wait_period);
+		}
+	}
+}
+
+static void kscan_ksi_isr_callback(const struct device *dev)
+{
+	k_sem_give(&poll_lock);
+}
+
+int kscan_matrix_init(const struct device *dev)
+{
+	struct kscan_matrix_data *const data = dev->data;
+
+	kscan_dev = DEVICE_DT_GET(KSCAN_DEV_NAME);
+	if (!device_is_ready(kscan_dev)) {
+		LOG_ERR("kscan device %s not ready", kscan_dev->name);
+		return -ENODEV;
+	}
+	/* Initialize semaphore usbed by kscan task and driver */
+	k_sem_init(&poll_lock, 0, 1);
+
+	/* Time figures are transformed from msec to usec */
+	data->deb_time_press = (uint32_t)(CONFIG_KSCAN_MATRIX_DEBOUNCE_DOWN * USEC_PER_MSEC);
+	data->deb_time_rel = (uint32_t)(CONFIG_KSCAN_MATRIX_DEBOUNCE_UP * USEC_PER_MSEC);
+	data->poll_period = (uint32_t)(CONFIG_KSCAN_MATRIX_POLL_PERIOD * USEC_PER_MSEC);
+	data->poll_timeout = 100 * USEC_PER_MSEC;
+
+	kscan_matrix_configure(kscan_dev, kscan_ksi_isr_callback);
+
+	k_thread_create(&data->thread, data->thread_stack, CONFIG_KSCAN_MATRIX_THREAD_STACK_SIZE,
+			(void (*)(void *, void *, void *))kscan_matrix_polling_thread, (void *)dev,
+			NULL, NULL, CONFIG_KSCAN_MATRIX_THREAD_PRIO, 0, K_NO_WAIT);
+
+	return 0;
+}
+
+static int kscan_matrix_config(const struct device *dev, kscan_callback_t callback)
+{
+	struct kscan_matrix_data *const data = dev->data;
+
+	if (!callback) {
+		return -EINVAL;
+	}
+	data->callback = callback;
+
+	return 0;
+}
+
+static int kscan_matrix_enable_interface(const struct device *dev)
+{
+	struct kscan_matrix_data *const data = dev->data;
+
+	atomic_set(&data->enable_scan, 1);
+
+	return 0;
+}
+
+static int kscan_matrix_disable_interface(const struct device *dev)
+{
+	struct kscan_matrix_data *const data = dev->data;
+
+	atomic_set(&data->enable_scan, 0);
+
+	return 0;
+}
+static const struct kscan_driver_api kscan_matrix_api = {
+	.config = kscan_matrix_config,
+	.enable_callback = kscan_matrix_enable_interface,
+	.disable_callback = kscan_matrix_disable_interface,
+};
+DEVICE_DEFINE(kscan_matrix, CONFIG_KSCAN_MATRIX_DRV_NAME, kscan_matrix_init, NULL, &kb_matrix_data,
```

Comment:
why not DT based devices?

Reply:
The kscan_matrix is a pseudo driver that registers with the existing kscan interface. It doesn't require bindings to the actual hardware peripheral.